### ISSUE TITLE
[CD] Enhancing the release process

### ISSUE THREAD

**Issue by cachedout** [elastic/apm-agent-java#991]
## Background

Currently the release process is a mix of [manual](https://github.com/elastic/apm-agent-java/blob/master/CONTRIBUTING.md#releasing) and [automated](https://internal-ci.elastic.co/view/APM/job/elastic+apm-agent-java+release/configure) steps[1]. We would like to have all APM agents move as closely as possible to a friction-free release process that's simple, reliable, and fast.

[1] [JJB for internal-ci](https://github.com/elastic/infra/tree/master/ci/jjb/internal-ci/defs/elastic%2Becs-logging-java)

## Proposal

What we want in an ideal world is for a single event to trigger a release and for that process to flow through the [APM Jenkins CI](https://apm-ci.elastic.co/), stopping only for cases where human confirmation is required. 

Below we break down how this process might look and what tasks might need to be completed in order to enable each. This completion of this project shouldn't be in a large PR but instead in incremental additions which can be built and tested independently before building up enough confidence to introduce automation across the entire release process.

## Release Process Steps

Below we detail each of the current twelve steps and how they might correspond to a step in a Jenkins pipeline.

_Questions are marked in italics_

**1. Check if Sonatype is up: status.maven.org**

Use the [HTTP Request Step](https://github.com/elastic/apm-pipeline-library/blob/master/vars/httpRequest.txt) to check status.maven.org. Fail the release process if it is not up.

**2. Review project version. The release version will be ${project.version} without the -SNAPSHOT. In case you want to update the version, execute mvn release:update-versions**

_Could somebody from the agent team clarify this requirement? In which cases does the project version differ from the release version?_

**3. Execute the release Jenkins job on the internal ci server.**

This will be done by duplicating what's in the internal Jenkins CI to the APM CI. There is [already a PR open to do this](https://github.com/elastic/apm-agent-java/pull/580).

**4. Login to oss.sonatype.org, go to Staging Repositories, close and release the staged artifacts.**

We will need to investigate the feasibility of automating this. [It appears as though this is well-supported on the Sonatype side](https://help.sonatype.com/repomanager3/rest-and-integration-api/script-api).

**5. Fetch and checkout the latest tag e.g. git fetch origin**

Easy enough. ;)

**6. If this was a major release, create a new branch for the major.**

Also easy enough.

**6.1 Add the new branch to the conf.yaml in the docs repo**

We could automate this but should we? This might be better left as a manual step, or perhaps as an artifact.

**7. If this was a minor release, reset the current major branch (1.x, 2.x etc) to point to the current tag, e.g. git branch -f 1.x v1.1.0**

Straightforward.

**8. Update CHANGELOG.asciidoc to reflect version release. Go over PRs or git log and add bug fixes and features.**

Not at all straightforward. Perhaps this should be considered a step to be completed before the release process is kicked off? Or should we consider fully automating the changelog as well? There are certainly strategies for doing so but is worth a chat with the team.

_Are there other teams doing this?_

**9. Go to elastic/apm-agent-java/releases and draft a new release. Provide a link to release notes in documentation as release description.**

Also a challenge to automate. :-/

**10. Update cloudfoundry/index.yml**

Should be reasonably easy to update automatically.

**11. Wait for released package to be available in maven central**

We'll probably have to write a script to do this?

**12. Publish release on Github. This will notify users watching repository.**

Should be trivial.

**Comment by felixbarny** [comment]
> 2. Review project version. The release version will be ${project.version} without the -SNAPSHOT. In case you want to update the version, execute mvn release:update-versions
> 
> Could somebody from the agent team clarify this requirement? In which cases does the project version differ from the release version?

When releasing a new version, for example `1.1.0`, the next version will be `1.1.1-SNAPSHOT`. But most of the time, the next version to release will be `1.2.0` unless we do a bugfix-only release which would be version `1.1.1`. Maybe we could tell the maven release plugin to default to creating a new minor instead of a new bugfix version. If we do want do release a bugfix version, we can create a branch. In this example we could branch off the `v1.1.0` tag and set the project version to `1.1.1-SNAPSHOT` so that a release would create the version `1.1.1`.

Another idea is to create a parameter for the release job so we have to type in the number of the version we want to release.

> 3. Execute the release Jenkins job on the internal ci server.
>
> This will be done by duplicating what's in the internal Jenkins CI to the APM CI. There is already a PR open to do this.

My concern with that is that the logs of the release process are public and on a publically accessible Jenkins instance. If there are any leaks of keys in the logs, they will leak to the public.

> 8. Update CHANGELOG.asciidoc to reflect version release. Go over PRs or git log and add bug fixes and features.
> 
> Not at all straightforward. Perhaps this should be considered a step to be completed before the release process is kicked off? Or should we consider fully automating the changelog as well? There are certainly strategies for doing so but is worth a chat with the team.
>
> Are there other teams doing this?

We already have a checkbox in the PR template to add a changelog entry. So this step is just to double-check that we didn't forget something. So yes, this can be considered a prerequisite.

**Comment by eyalkoren** [comment]
> 6.1 Add the new branch to the conf.yaml in the docs repo
>
> We could automate this but should we? This might be better left as a manual step, or perhaps as an artifact.

Major releases are very infrequent, so the ROI on automating that would be very low. Besides, I assume we would want such addition to the docs to be tested manually.

**Comment by cachedout** [comment]
> My concern with that is that the logs of the release process are public and on a publically accessible Jenkins instance. If there are any leaks of keys in the logs, they will leak to the public.

Hi @felixbarny. This is indeed a concern but we are already handling masking of sensitive data in quite a number of places on `apm-ci` presently.

I see two main advantages to moving the release process automation to `apm-ci`.

1. We would like to have a single pipeline that can run through build and test and then into the release stage if desired. Having the testing/CI process on one server and the release process on another breaks the goal of a unified automated workflow that we would like to achieve.

2. Having the pipeline configuration for the release process in this repo instead of in the internal repo where it is now stored means that we could have everything in one place, which is nice.

That said, of course, if you're worried about moving to a public CI we should definitely keep discussing it. :) Thanks!

cc: @elastic/observablt-robots

**Comment by felixbarny** [comment]
I'm definitely not completely against it and we have already been bitten by different configurations on release CI vs the normal CI so I'd welcome this if we can make sure that we don't leak keys etc.

Not suggesting we should move to it but did you consider GitHub workflows? Just wondering if you have experience or thoughts on it.

**Comment by cachedout** [comment]
I have spent some more time today looking at this and I have a few questions for you, @felixbarny 

Are any of the steps defined here more painful than others? If not, is there a place you'd like to see efforts focused? 

Are there steps you'd feel uncomfortable with us trying to automate?

Happy to schedule a quick call to go over it but I thought I'd pose these questions first and see what comes back. :)

**Comment by felixbarny** [comment]
The steps in isolation are all are not that painful. It's just the amount of steps that make the release process lengthy and error-prone. So I'd start with whatever is easiest to automate.

**Comment by cachedout** [comment]
### Staging to Release process on Sonatype

Right now, what happens is that artifacts are uploaded to Sonatype automatically during the CI build. The command which is ultimately run to achieve this is `./mvnw release:prepare release:perform --settings .ci/settings.xml -Darguments="--settings .ci/settings.xml" --batch-mode`.

With the `prepare` step, all the `pom.xml` files are rewritten from their current state (most likely x.x.x-SNAPSHOT) to a semver-complient (x.x.x).

With the `perform` step, artifacts are actually uploaded to Sonatype and placed into the staging area which is located at https://oss.sonatype.org/service/local/staging/deploy/maven2/

While the non-paid version of Sonatype [does not offer staging capability](https://help.sonatype.com/repomanager3/staging), which makes local testing difficult, it does appear to be present on `oss.sonatype.org`.

### Proposal

We handle promotion by simply using the [promotion endpoint](https://help.sonatype.com/repomanager3/staging#Staging-Promotion). This should be a single call and very easy. :)

### Questions

It would be nice if we could have artifacts which are uploaded to staging tagged automatically. That would give us a definitive target to use when constructing the URL to use to call to the promotion endpoint.

One way to tag artifacts automatically might be to use the [NXRM3 Maven Plugin](https://github.com/sonatype/nxrm3-maven-plugin). I don't know enough about this to know if this is a good idea or even feasible, so I'm hoping to get feedback from @eyalkoren and @felixbarny before moving forward. Concretely, we'd have to have to modify the `pom.xml` file to support this and perhaps there are other changes that I'm not aware of? I could use some guidance on this point.

If we can tag releases as they are uploaded, then promotion should be a breeze.

If we don't use tags, it seems like we should still be able to construct a search query to target the staged artifacts, but we'll have to play around with that. 

### Next steps

Waiting for feedback from @eyalkoren and @felixbarny on possibly tagging during the initial upload to staging or whether we should go with an alternative.

**Comment by eyalkoren** [comment]
I don't know about tagging either. 
I briefly looked into it and saw that it makes use of tags, but didn't see how it can be used to create tags. In any case, my gut feeling is that minimising the poms' involvement in these post-build stuff is a good practice, unless it makes things much easier.
What about using [the Add Tag endpoint](https://help.sonatype.com/repomanager3/tagging#Tagging-AddTag)? Can this be used or does this brings the same obstacle of finding proper `attributes` to identify the right artefacts?

**Comment by cachedout** [comment]
We should be fine to go with a search-based approach instead of using tag since it looks like there is just one artifact that we're dealing with. I'll experiment with this now and will ensure that it will work well.

**Comment by cachedout** [comment]
I've spent the morning learning a bit more than I ever wanted to know about Nexus repository management and I'll put my findings below for comment.

### tl;dr

While it remains desirable, it does not appear to be feasible to create automation for closing and promoting staged artifacts that lives independently of the build steps which are currently contained in `pom.xml`.

### oss.sonatype.org is v2.x and not v3.x

[Earlier](https://github.com/elastic/apm-agent-java/issues/991#issuecomment-579792033), I mentioned that we may be able to use the API endpoint provided by Sonatype for `staging/`. Unfortunately, this is was introduced in their 3.x series of Nexus. [The OSS repo is Nexus Repository Manager 2.14.15-01](https://oss.sonatype.org). Therefore, the needed endpoint is not there to call.

The 2.x version of the repository manager [does have an API](https://repository.sonatype.org/nexus-restlet1x-plugin/default/docs/index.html) but it does not appear to contain needed elements, such as an interface to staging repos.

There are two approaches which seem viable. I'll explain both below.

#### Jenkins option

I investigated the feasibility of using the Sonatype-provided Jenkins plugin. While it has a number of interesting features, the staging features are all limited to 3.x Nexus installations. Therefore, I ruled the use of it out for the purposes of this project, though it remains an interesting option for cases down the road where we might have the chance to use a 3.x Nexus repo.

#### Maven plugin option

This appears to be the officially sanctioned approach for solving the problem of dealing with staging repositories in Nexus 2.x. 

Specifically, in the documentation describing the best practices for dealing with their OSS repo, [we find this](https://central.sonatype.org/pages/releasing-the-deployment.html):

`Note that deploying with the Nexus Staging Maven plugin or Ant tasks will, by default, automatically attempt to close the staging repository upon deployment and can be used to release the staging repository from the command line as well. This allows you to avoid logging into the OSSRH user interface altogether. More details can be found in the Maven and Ant sections of this guide.
`

In the Maven guide referenced above, [we find their statement on interacting with the OSS repo programatically](https://central.sonatype.org/pages/apache-maven.html):

`The Nexus Staging Maven Plugin is the recommended way to deploy your components to OSSRH and release them to the Central Repository. `

#### Conclusion

I know that it's not necessarily our first choice, but given the constraints that we're facing which are out of our hands with the version of the remote API on the Nexus side, I think our best path forward is to add a new plugin to the project's `pom.xml' [as described in their documentation](https://central.sonatype.org/pages/apache-maven.html) under the section *Nexus Staging Maven Plugin for Deployment and Release* which would look approximately like the following:

```
<plugin>
  <groupId>org.sonatype.plugins</groupId>
  <artifactId>nexus-staging-maven-plugin</artifactId>
  <version>1.6.7</version>
  <extensions>true</extensions>
  <configuration>
     <serverId>ossrh</serverId>
     <nexusUrl>https://oss.sonatype.org/</nexusUrl>
     <autoReleaseAfterClose>true</autoReleaseAfterClose>
  </configuration>
</plugin>
```

This should let us simply do a `mvn nexus-staging:release` as a part of the CD pipeline.

#### Next steps

I know this is something that the dev team will want to consider since it's a change to the project's `pom.xml` so I will wait to hear back from @eyalkoren or @felixbarny. Thanks!

**Comment by eyalkoren** [comment]
@elastic/observablt-robots Since our release process is so tedious, we tend to release not so frequently. Do you think you can allocate some resources to try and work on this enhancement?

**Comment by cachedout** [comment]
@eyalkoren I have actually been working on it for the past several days. :) I am very close to finishing all the needed code to interact programmatically via our Jenkins Pipelines with the Sonatype repository. From there, we'll be able to put the rest of the pieces for automation into place quite easily.

**Comment by eyalkoren** [comment]
That would be AWESOME, Thanks!
